File: frontend/src/utils/reservedReadmeTitles.ts
/**
 * Note titles reserved for notebook/folder readme content (not ordinary notes).
 * Matches `readme` and `readme.md` case-insensitively — keep in sync with backend
 * `ReservedReadmeTitles`.
 */
export const RESERVED_README_TITLE_MESSAGE =
  "'readme' and 'readme.md' are reserved for notebook and folder readme content."

export function isReservedReadmeNoteTitle(
  title: string | null | undefined
): boolean {
  if (title == null) return false
  const trimmed = title.trim().toLowerCase()
  return trimmed === "readme" || trimmed === "readme.md"
}

File: backend/src/main/java/com/odde/doughnut/validators/ReservedReadmeTitles.java
package com.odde.doughnut.validators;

/**
 * Note titles reserved for notebook/folder readme content (not ordinary notes). Matches {@code
 * readme} and {@code readme.md} case-insensitively.
 */
public final class ReservedReadmeTitles {

  public static final String RESERVED_MESSAGE =
      "'readme' and 'readme.md' are reserved for notebook and folder readme content.";

  private ReservedReadmeTitles() {}

  public static boolean isReserved(String title) {
    if (title == null) {
      return false;
    }
    String trimmed = title.trim();
    return trimmed.equalsIgnoreCase("readme") || trimmed.equalsIgnoreCase("readme.md");
  }
}
